library/src/main/kotlin/org/codice/compliance/report/Report.kt

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright (c) 2018 Codice Foundation
+
+Released under the GNU Lesser General Public License; see
+http://www.gnu.org/licenses/lgpl.html
+*/
+package org.codice.compliance.report
+
+import org.codice.compliance.QUIET_MODE
+import org.codice.compliance.SAMLComplianceException
+import org.codice.compliance.report.Report.Section.GENERAL
+import org.codice.compliance.report.Report.Section.SCHEMA
+import org.fusesource.jansi.Ansi
+import java.io.File
+import java.io.PrintWriter
+
+object Report {
+
+    internal const val REPORT_FILE = "report.txt"
+    private const val NOT_TESTED_LIST =
+            "https://github.com/codice/saml-conformance/blob/master/ctk/idp/NotTested.md"
+
+    private val emptyExceptionMap: MutableMap<Section, MutableSet<SAMLComplianceException>?> =
+            Section.values().map {
+                it to null
+            }.toMap().toMutableMap()
+
+    private val exceptionMessages = mutableMapOf<Section, MutableSet<SAMLComplianceException>?>()
+            .apply {
+                putAll(emptyExceptionMap)
+            }
+
+    private var currentTestExceptionMessages = mutableSetOf<SAMLComplianceException>()
+
+    private val reportQuietly by lazy {
+        QUIET_MODE.toBoolean()
+    }
+
+    private var hasExceptions = false
+
+    /**
+     * Adds an exception to the {@code exceptionMessages} for the given {@param exception}'s section
+     *
+     * @param exception - the SAML Compliance Exception to add
+     */
+    fun addExceptionMessage(exception: SAMLComplianceException): Report {
+        val section = exception.section
+        if (!section.isStarted()) {
```

Comment:
✏️ I believe you can collapse these two overloaded methods into one (`exception` has to be declared first in the signature to support this):
```kotlin
fun addExceptionMessage(exception: SAMLComplianceException, section: Section = exception.section): Report {
        if (!section.isStarted()) {
            exceptionMessages[section] = mutableSetOf(exception)
        } else {
            exceptionMessages[section]?.add(exception)
        }

        currentTestExceptionMessages.add(exception)
        hasExceptions = true
        return this
    }
```
Even if that doesn't work, you can definitely invoke the overloaded method after you've extracted the `section` here.

Reply:
That will work. We just use the "explicitly give the section" part in like two places